keps/sig-node/1948-add-deallocate-to-device-plugin-api/README.md

```diff
@@ -100,12 +100,10 @@ Only risk is breaking existing device plugins by introducing non optional change
 
 ## Design Details
 
-- Remove reusable devices from `devicemanage`. Reusable devices are silently reused by kubernetes, and so they can't be reported back to device plugins for cleanup, tracking, etc.
 - Move `PreStartContainer` in `DeviceManager` to be used as a container lifecycle hook. (This change isn't truly required, but useful for compatibility and organization with the next steps).
 - Add `PostStopContainer` and `Deallocate` calls in the DevicePlugin API.
 - Add `PostStopContainer` as a container lifecycle hook.
-- Add `Deallocate` calls in container manager.
-- Delete test cases for reusable devices.
+- Add `Deallocate` calls in container manager, taking care to only do so for devices that are no longer in the reuse list.
```

Comment:
This will make it asymmetrical with the `Allocate()` call. For any given device, `Allocate()` will be called for all init-containers, plus the final app-container, but `Deallocate()` will only be called after the final app-container is done with it. This feels awkward.

If we do go this route, I'd say we also need to do this for `PostStopContainer()` as well, as we don't want any logic here to de-initialize a device, if the whole point of running it through an init-container was to initialize it in the first place. This is complicated by the fact that the `PreStopContainer()` already has some strange semantics in that it will be called on all init/app containers which may undo work (or do redundant work) of previous containers in the chain.

I'm wondering if the right answer here is actually to keep both of these symmetric with their `Allocate()` and `PreStartContainer()` counterparts, but add a field to all 4 calls indicating whether the container allocating/deallocating the devices is an init container or an app container.

I imagine the plugin to then do something different depending on which type of container the call is coming in for.

Reply:
@klueska +1 for keeping both APIs in sync with their counterparts and adding parameter for all 4 calls.
@mewais wdyt? Would you be willing to update the PR?

Reply:
@bart0sh I agree, will update this as soon as I can.

Reply:
@mewais any updates? I can help if needed.

Reply:
@klueska   
```
commit 0b168f0243660a4b5f5c65d2b76ce828362ffa88
Author: Kevin Klues <kklues@nvidia.com>
Date:   Sun Feb 2 22:00:24 2020 +0000
```
Hi, excuse me. Could you please explain a bit why we need `to allocate devices on a container-by-container basis`?  
I've thought about this for a while, but got no good answer...  

If the interfaces were 'pod' level as them used to be, we can just take the pod as a whole, we don't have to take 'init' or 'app' containers differently.

Reply:
And maybe we have other approaches to resolve the difference between `init` and `app` containers.

Just for an example, we may have two 'post-stop' interfaces, `PostGracefulStopContainer` for the container exits by itself, and `PostEnforcedStopContainer` for the container gets closed.  
Normally, init-container should finish its job and exit by itself. In this case, `PostGracefulStopContainer` is called, which does not undo init-container's work.  
If init-container fails to finish its job, may be interrupted in the middle, the pod would be restarted soon. So it is okay to undo its work in `PostEnforcedStopContainer`.  
For app-container, the same logic can be applied. One of these two interfaces is triggered when container's lifecycle ends.

Reply:
@klueska friendly ping, could you check the previous comments from @Windrow14 ?

Reply:
> +1 for keeping both APIs in sync with their counterparts and adding parameter for all 4 calls.

👍 this makes sense to me to avoid confusion and complication.

@mewais is not responding, how can we proceed this?

Reply:
@fujitatomoya
> @mewais is not responding, how can we proceed this?

You can create your own PR and continue with it I guess.

Reply:
@FengGaoCSC @Windrow14 could you create PR with `Co-authored` @mewais ?

Reply:
@fujitatomoya https://github.com/kubernetes/enhancements/pull/3074

I created a new PR, is it in good shape?  
Excuse me, but how to set `co-authored`? Is it a button or a flag?